File: rest-service/src/main/java/mb/lib/blast/model/n/BlastNTask.java
package mb.lib.blast.model.n;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import mb.api.service.util.ErrorText;

public enum BlastNTask
{
  BlastN("blastn"),
  BlastNShort("blastn-short"),
  DiscontiguousMegablast("dc-megablast"),
  Megablast("megablast"),
  RMBlastN("rmblastn");

  private final String value;

  BlastNTask(String value) {
    this.value = value;
  }

  @JsonValue
  public String getValue() {
    return value;
  }

  @JsonCreator
  public static BlastNTask fromString(String value) {
    for (var e : values())
      if (e.value.equals(value))
        return e;

    throw new IllegalArgumentException(String.format(
      ErrorText.InvalidEnumValue,
      value,
      BlastNTask.class.getSimpleName()
    ));
  }

  @Override
  public String toString() {
    return value;
  }
}

File: lib-jvm-mblast-sdk/src/main/kotlin/org/veupathdb/lib/mblast/sdk/query/blast/BlastNTask.kt
package org.veupathdb.lib.mblast.sdk.query.blast

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.TextNode
import org.veupathdb.lib.mblast.sdk.util.UnrecognizedEnumException
import org.veupathdb.lib.mblast.sdk.util.dEnumRequireTextual

enum class BlastNTask {
  BlastN,
  BlastNShort,
  DCMegaBlast,
  MegaBlast,
  RMBlastN,
  ;

  @JsonValue
  fun toJson() = TextNode(
    when (this) {
      BlastN      -> "blastn"
      BlastNShort -> "blastn-short"
      DCMegaBlast -> "dc-megablast"
      MegaBlast   -> "megablast"
      RMBlastN    -> "rmblastn"
    }
  )

  companion object {

    @JvmStatic
    @JsonCreator
    fun fromJson(js: JsonNode): BlastNTask {
      js.dEnumRequireTextual(BlastNTask::class)

      return when (val v = js.textValue()) {
        "blastn"       -> BlastN
        "blastn-short" -> BlastNShort
        "dc-megablast" -> DCMegaBlast
        "megablast"    -> MegaBlast
        "rmblastn"     -> RMBlastN
        else           -> throw UnrecognizedEnumException(BlastNTask::class, v)
      }
    }
  }
}